Insights data instrumentation

paolodamico opened the pull request:
## Changes

Adds some key instrumentation related to insights.
- Introduces `insight viewed` event which is triggered when a user visits an insights graph (trends, sessions, retention ...).
    - The event is also triggered anytime the filtering properties change.
    - The event is de-bounced for 500ms. This is just to avoid noisy events from a user changing the filters multiple times in a row.
    - The event supports both states for the `remove-shownas` feature flag introduced on #2899. Regardless of the state of the feature flag, the `insight` property will be set to `TRENDS`, `LIFECYCLE` or `STICKINESS` as appropriate.
- Introduces `dashboard viewed` event which is triggered when a user visits a dashboard.

Please review the code for context on all the included metadata on each tracked event.

**Jan 12 update**
Per discussion with @macobo & @jamesefhawkins, we've removed the requirement for the user to stay in the page for 5s before firing the event. This to avoid adding complexity that brings little value, and the 5s rule can be arbitrary (i.e. value can be driven by looking at a graph for less than 5 seconds and conversely, staying 5s does not guarantee driving value).

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

paolodamico committed:
insight viewed event

paolodamico committed:
abstract reportUsage, debounce report after 5secs

paolodamico committed:
added actions count

paolodamico committed:
handle edge case for trends

paolodamico committed:
viewed dashboard event

paolodamico committed:
add watching_shared prop & debounce

paolodamico commented:
Tagging @jamesefhawkins in case you have any comments on the metadata we're attaching.

To whomever reviews the code: as this will be used for key metrics, can you help me dive into the weeds to be as certain as possible that we're capturing these events correctly?

Hopefully we can merge on cloud before the final code freeze.

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -43,6 +46,66 @@ export const insightLogic = kea<insightLogicType>({
+            actions.reportUsage(filters.filters)
+            actions.setNotFirstLoad()
+        },
+        reportUsage: async (filters: Record<string, any>, breakpoint) => {
```
What's the real type for filters here? Since it only seems to have 4 valid keys, can we create a proper interface for it?

Without doing so, all of the keys below will have type `any` and we're basically not making use of typescript.

We can't really make use of exhaustive pattern matching for `insight` type as well - it's possible to write this code in a way that if a new view is added, this place will start to error and we'll know to add another case to the `if` below.

Happy to pair on this.

paolodamico replied:
I think it's a bit trickier, filters can really be string, number, an object, an array of strings/numbers/objects. They also seem to have a ton more keys. 100% agree we should type this better, but might be outside the scope of this? Will ping you to see if it's worth fixing rn.

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -43,6 +46,66 @@ export const insightLogic = kea<insightLogicType>({
+            }
+
+            const sanitizedProperties: Record<string, any> = {}
+            Object.entries(properties).map(([key, value]) => {
```
Undefined properties already get dropped (this is default JSON.stringify behavior) - this is dead code.

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -43,6 +46,66 @@ export const insightLogic = kea<insightLogicType>({
+                date_from,
+                date_to,
+                filters_count: filters.properties ? filters.properties.length : 0, // Only counts general filters (i.e. not per-event filters)
+                events_count: filters.events ? filters.events.length : undefined, // Number of event lines in insights graph; number of steps in funnel
```
Can be simplified: `filters.events?.length`

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -43,6 +46,66 @@ export const insightLogic = kea<insightLogicType>({
+        reportUsage: async (filters: Record<string, any>, breakpoint) => {
+            // Reports `insight viewed` event
+            const { insight, display, interval, date_from, date_to } = filters
+            const properties: Record<string, any> = {
```
this is `Record<string, string | number` right? :)

paolodamico replied:
Will continue discussion on https://github.com/PostHog/posthog/pull/2787#discussion_r544891453

macobo commented on frontend/src/scenes/dashboard/dashboardLogic.js:
```diff
@@ -399,5 +398,24 @@ export const dashboardLogic = kea({
                 setTimeout(() => actions.refreshDashboardItem(id), 1000)
             }
         },
+        reportUsage: async (payload, breakpoint) => {
```
What's the type for payload? I assume it's defaulting to `any` right now.

paolodamico replied:
Well this file is still JS so no typing 😢

macobo replied:
aha :D nvm then

macobo commented on frontend/src/scenes/dashboard/dashboardLogic.js:
```diff
@@ -399,5 +398,24 @@ export const dashboardLogic = kea({
+            const { created_at, name, is_shared, pinned } = payload
+            const properties = { created_at, name, is_shared, pinned, sample_items_count: 0 }
+            properties.item_count = payload.items.length
+            properties.created_by_system = payload.created_by ? false : true
```
```js
const properties = {
  created_at,
  name,
  is_shared,
  pinned,
  sample_items_count: 0,
  item_count: payload.items.length,
  created_by_system: !payload.created_by,
  watching_shared: !!props.shareToken
}
```

macobo reviewed (commented):
Noticed some problems re ts - happy to help iron these out. :)

Change is :+1:

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -43,6 +46,66 @@ export const insightLogic = kea<insightLogicType>({
+        },
+        reportUsage: async (filters: Record<string, any>, breakpoint) => {
+            // Reports `insight viewed` event
+            const { insight, display, interval, date_from, date_to } = filters
```
```js
interface Foo {
  insight: 'TRENDS' | '...',
  display?: string
  interval?: string
  date_from?: string
  date_to?: string 
}
```

paolodamico committed:
fix some ts + address feedback

paolodamico committed:
fix retention edge case

paolodamico commented:
Thanks for the thorough review @macobo addressed all your feedback. Re filters typing, agreed on taking small steps towards improving typing here, starting with this new action payload! As we make sure it captures all cases we can use this same interface for the general filters.

Ready for a quick final look if you want too. I also forgot to ask if you got a chance to actually test the functionality and make sure we're triggering the events at the right time.

EDsCODE commented:
Also heads up, typing the filters is on mine and Tim's release cycle. We haven't prioritized it so far because we've been doing more of the backend fixes but can pick up whever you leave it

paolodamico commented:
Re @EDsCODE good to know! Hope this gives you a bit of a head start

macobo commented:
Thanks for doing the changes! Looking great, I wouldn't mind if you just went ahead and merged it. I did not do any manual testing (yet) - can do so tomorrow if you want to wait!

paolodamico commented:
Thanks @macobo! I actually prefer to wait to make sure this is as accurate as possible as it'll be used for key metrics. I don't think we should merge anything new anyways.

paolodamico committed:
adjust breakpoints, simplify definition

paolodamico committed:
Merge branch 'master' into instrumentation-d15

paolodamico committed:
Merge branch 'master' into instrumentation-d15

paolodamico committed:
add support for 2899

paolodamico commented:
@macobo ready for another view, I've updated according to your comments on #2811 and what we discussed yesterday. PR description updated too.

If you could help testing too, it'd be incredibly helpful.

macobo commented on frontend/src/scenes/dashboard/dashboardLogic.js:
```diff
@@ -319,5 +318,30 @@ export const dashboardLogic = kea({
+                watching_shared: !!props.shareToken,
+            }
+
+            for (const item of payload.items) {
```
Not-actionable: We're adding a ton of cruft like this for analytics everywhere - the extra lines of code do make things less maintainable as it's not always clear what's real code and what's analytics.

If we were using lodash, we could get rid of most lines of code using it here using sums/groupBys. The arrays here are small enough that there isn't really a performance win writing long loops.

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -11,6 +12,32 @@ export const ViewType = {
+    interval?: string
+    date_from?: string
+    date_to?: string
+    properties?: Record<string, any>[]
```
Note, not actionable: `any` is the worst type. :)

macobo replied:
This type already exists: PropertyFilter from types.ts

So this should be `properties: Array<PropertyFilter>`

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -11,6 +12,32 @@ export const ViewType = {
+    date_from?: string
+    date_to?: string
+    properties?: Record<string, any>[]
+    events?: Record<string, any>[]
```
Might be type Entity from types.ts, not 100% sure

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -11,6 +12,32 @@ export const ViewType = {
+    session?: string
+    period?: string
+    retentionType?: 'retention_recurring' | 'retention_first_time'
+    returningEntity?: Record<string, any>
```
Entity from types.ts

macobo commented on frontend/src/scenes/insights/insightLogic.ts:
```diff
@@ -45,6 +74,77 @@ export const insightLogic = kea<insightLogicType>({
+            properties.total_event_actions_count = (properties.events_count || 0) + (properties.actions_count || 0)
+
+            // Custom properties for each insight
+            if (insight === 'TRENDS') {
```
Not blocking: 

So this event is going to be hard to analyze because of properties sometimes being there and other times not.

On the other end, it's also a ton of code we now need to maintain and might start blowing up any time we change anything about filters leading to missing metrics.

Would we achieve the same with `properties.filters = filters`?

paolodamico replied:
As discussed on Slack, we'll keep both for now as having `filters` can future-proof us better, but having the parsed version enable us to do analytics **today** (with the disclaimer that this code may be a bit fragile).

EDsCODE reviewed (commented):
Karl caught many of the details. My thoughts resulting from his comments are that it might be worth having an individual logic for analytic events? The reportusage action can just exist separately and be injected when needed. This will reduce the confusion of app code and analytics code.

paolodamico committed:
Merge branch 'master' into instrumentation-d15

paolodamico committed:
refactor reportInsightViewed to eventUsageLogic

paolodamico committed:
move reportViewedDashboard to eventUsageLogic

paolodamico commented:
Thanks for the feedback, I've addressed all of it and also per @EDsCODE's suggestion, moved all the analytics logic to the `eventUsageLogic` component.
